frontend/src/redux/middleware.js

```diff
@@ -0,0 +1,55 @@
+import { addWidget } from '@metacell/geppetto-meta-client/common/layout/actions';
+import { WidgetStatus } from '@metacell/geppetto-meta-client/common/layout/model';
+import { v4 as uuidv4 } from 'uuid';
+import { ADD_VIEWER } from './actions/viewers';
+import { ADD_DEVSTAGE } from './actions/devStages';
+import { raiseError, waitData } from './actions/misc';
+import { DevStageService } from '../services/DevStageService';
+
+const devStagesService = new DevStageService();
+
+const middleware = (store) => (next) => (action) => {
+  function widgetFromViewerSpec(viewerSpec) {
+    return {
+      id: viewerSpec.viewerId,
+      name: `${viewerSpec.type}_${viewerSpec.viewerId}`,
+      component: viewerSpec.type,
+      panelName: 'centralPanel',
+      enableClose: true,
+      enableRename: true,
+      enableDrag: true,
+      status: WidgetStatus.ACTIVE,
+      config: {
+        viewerId: viewerSpec.viewerId,
+      },
+    };
+  }
+
+  switch (action.type) {
+    case ADD_DEVSTAGE: {
+      next(waitData('Getting development stages...', action.type));
+      devStagesService.getDevStages().then((stages) => {
+        // eslint-disable-next-line no-param-reassign
+        action.data = stages;
+        next(action);
+      }, (e) => {
+        // eslint-disable-next-line no-console
+        console.error('Error getting development stages', e);
+        next(raiseError('Error getting development stages'));
+      });
+      break;
+    }
+    case ADD_VIEWER: {
+      // eslint-disable-next-line no-param-reassign
+      action.data.viewerId = uuidv4();
```

Comment:
don't change the action here, rather create a new action object based on this one

`const viewerSpec = {...action, viewerId: uuidv4()}`